### ISSUE TITLE
VSCodeDebugProtocol returns lowercase breakpoint paths

### ISSUE THREAD

**Issue by AndrewNatoli** [Microsoft/vscode#1916]
VSCode 0.10.6 on Mac OSX 10.11.1
Using [vscode-php-debug](https://github.com/felixfbecker/vscode-php-debug) (XDebug adapter for PHP debugging)

------

`VSCodeDebugProtocol.SetBreakpointsArguments()` returns lowercase paths for breakpoints. 

XDebug expects to see a path like `/Developer/Project/file.php` but we're setting a breakpoint for `/developer/project/file.php`. This has allegedly been [resolved with an option in XDebug 2.3](http://bugs.xdebug.org/view.php?id=973) but as per the nature of that release, developers working on PHP <= 5.3 projects do not benefit from that fix.

I made a [proof of concept change](https://github.com/AndrewNatoli/vscode-php-debug/commit/d0c7569e5d9c12830ab9f9e0fc499b944ec1415e) that replaces the lowercase path returned by `VSCodeDebugProtocol.SetBreakpointsArguments()` with the actual `cwd` and it resolved the issue. Alas, if it's possible without having adverse effects elsewhere, it would make sense to return the correct path to begin with instead of needing each adapter to program around this. 

Original report: https://github.com/felixfbecker/vscode-php-debug/issues/8

**Comment by weinand** [comment]
@AndrewNatoli I've just verified that VSCode passes file paths unmodified to the debug adapter trough the SetBreakpointsArguments.source.path. So the path that you see in the debug adapter should match the path of the file on disk (If this would not be the case, debug adapters would not work on case sensitive file systems e.g. linux).

**Comment by AndrewNatoli** [comment]
Hi @weinand,
Thanks for checking. I'm still getting lowercase paths so I'll try digging further and trying out other debug adapters then get back to you on this.

**Comment by weinand** [comment]
@AndrewNatoli Are you saying that if you set a breakpoint in a file "/Developer/Project/File.php" VSCode sends you the path "/developer/project/file.php" in the SetBreakpointsRequest?

**Comment by AndrewNatoli** [comment]
@weinand Yes, that's correct

**Comment by weinand** [comment]
@isidorn Is this possible?

**Comment by isidorn** [comment]
I do not see that on the latest vscode on my mac

![screen shot 2016-01-11 at 15 52 19](https://cloud.githubusercontent.com/assets/1926584/12236659/67543422-b87b-11e5-8db4-79660245dbd5.png)

**Comment by AndrewNatoli** [comment]
Are the breakpoints usually sent through any custom code before `setBreakPointsRequest`? 

Here's my `source.path` immediately after setting into `setBreakPointsRequest`.

![screen shot 2016-01-11 at 11 04 12 am](https://cloud.githubusercontent.com/assets/4162073/12238660/1d9cb2aa-b853-11e5-8402-60b86c24d717.png)

My first directory should be "Developer," not "developer."

**Comment by weinand** [comment]
@AndrewNatoli this doesn't look like an OS X path since it starts with two slashes. What is your VS Code setup? Is this some shared drive?

**Comment by AndrewNatoli** [comment]
@weinand VS Code is installed on my primary (and only) drive at `/Applications/Visual Studio Code.app`

**Comment by weinand** [comment]
@AndrewNatoli and you've launched VS Code by double clicking the app in /Applications? And what is /Developer? This looks non-standard. I don't have such a root level folder. Could you try to use a php project that lives in your home directory? Does this makes a difference?

**Comment by weinand** [comment]
@AndrewNatoli your primary drive is not case-sensitive, right?

**Comment by AndrewNatoli** [comment]
@weinand Right, my primary drive it not case-sensitive.

/Developer is a leftover artifact from before XCode 4.3. I moved in and made it my document root when I was starting off.

I moved the project to my home directory.

I'm running VS Code by opening the app in /Application then testing the adapter by running

`[path to vscode.app]/Contents/MacOS/Electron /Users/andrew/vscode-php-debug/testproject --extensionDevelopmentPath=/Users/andrew/vscode-php-debug/`

Where /testproject is a sample php application. 

Here's an interesting development. When I add a breakpoint to the PHP project, `setBreakPointsRequest` is hit with the lowercase and double-slashed path (//users/andrew...).

Now, in Apache (I'm using MAMP in this scenario), if I make the default host's document root `/Users/andrew/vscode-php-debug/` and run a script with a breakpoint in my testproject directory, `setBreakPointsRequest` is hit again but this time with the correct path.

NOW, if I use an Apache virtual host and instead of accessing my app via `localhost` I use something like `test.mygreatapp.com`, when I run the script I get the lowercase, double-slashed path.

![screen shot 2016-01-11 at 11 58 33 am](https://cloud.githubusercontent.com/assets/4162073/12240418/4810adb8-b85b-11e5-8077-7057591c25c2.png)

![screen shot 2016-01-11 at 12 07 27 pm](https://cloud.githubusercontent.com/assets/4162073/12240523/eab3ef62-b85b-11e5-9ddf-64eb16349d11.png)

This occurs both in my home directory and my /Developer directory.

**Comment by weinand** [comment]
@AndrewNatoli I've tried to reproduce your setup with VSCode 0.10.6 but I'm not seeing the lower case paths:
![2016-01-11 23-04-31](https://cloud.githubusercontent.com/assets/1898161/12248175/f258a666-b8b7-11e5-9026-2b35534dfb87.png)
The only difference in my setup: I did not install anything php related (so I don't have xdebug) and I don't have Apache.

But I found this: http://stackoverflow.com/questions/7699412/case-inconsistency-of-php-file-paths-on-mac-mamp 

Is the problem introduced by Apache/MAMP?

**Comment by AndrewNatoli** [comment]
Hi @weinand,

It seems to be isolated to using a virtual host with Apache. By that I mean configuring Apache to route custom hostnames like `mygreatsite.com/testproject/app.php` instead of `localhost/testproject/app.php`. I'm not sure how this could possibly have an effect on what VSCode is seeing but the difference is there.

When using XDebug, XDebug sends an "init" message that includes the server-side path of the file you are running. So it will say, "Hey, Apache and I are serving up /Users/andrew/test.php."

The adapter will then say "Hey, that's cool XDebug. Listen, my guy set a breakpoint on /users/andrew/test.php. Could you stop there if you open that file?"

So in short, the Apache side of things seem fine while VSCode's output is incorrect. 

I'll try testing with a remote host in a moment.

------

I checked the bug report linked on that Stack Overflow post. Running their [sample code](https://bugs.php.net/bug.php?id=60017) did not produce the same result; the proper case is presented in paths.

**Comment by weinand** [comment]
@AndrewNatoli How can Apache affect VSCode? I could only imagine that some URL conversion that VSCode uses internally is affected by something like the virtual host feature of Apache. If this is the case, it will affect all debug adapters. Could you try whether you see the same issue with vscode-mock-debug?

**Comment by weinand** [comment]
@isidorn what code are you using to convert VSCode's internal URLs to the absolute pathnames sent to the debug adapter? Could you please help @AndrewNatoli to debug the client side of this so that we understand where paths become broken.

**Comment by felixfbecker** [comment]
I would not exclude the possibility of me screwing up the paths while converting them to URIs, so the XDebug logs are not 100% credible. But the screenshots @AndrewNatoli provided indicate that the arguments from VS Code (`args.source.path`) are indeed lowercased. @AndrewNatoli When you debug the adapter, all communications between Code and the adapter are logged to the console, are the paths lowercased there? That would be evidence for me.

And what file system are you using?

**Comment by weinand** [comment]
@AndrewNatoli can you please confirm that VSCode passes lower-cased paths to the php debug adapter even if XDebug is not running and Apache was not used to hit the code.

**Comment by AndrewNatoli** [comment]
Hey @weinand, I'll review and get back to you when I have a moment to check.

**Comment by weinand** [comment]
@AndrewNatoli great, thanks a lot!

**Comment by AndrewNatoli** [comment]
Hey @weinand,
`args.source.path` in `setBreakPointsRequest` comes in properly when setting a breakpoint.

**Comment by AndrewNatoli** [comment]
Just some further insight: while I can't explain how in the adapter the breakpoint paths were being translated to lowercase, I have yet to be able to reproduce this issue since the VSCode-to-XDebug communication has been fixed in the adapter. 

I'll update this if I run into an issue again while using the adapter during my usual weekday projects, otherwise I believe this is safe to close for now?

**Comment by weinand** [comment]
@AndrewNatoli Thanks for looking into this again and closing it.
